### ISSUE TITLE
Server Hibernation/Suspend

### ISSUE THREAD

**Issue by 0xCyberFortress** [pterodactyl/panel#4769]
### Is there an existing feature request for this?

- [X] I have searched the existing issues before opening this feature request.

### Describe the feature you would like to see.

Please don't just close this without reading, I'm looking for genuine solutions and hope what I have to offer can actually help.

I'm looking for a way to have servers auto-suspend or hibernate (not entirely shut down, if possible) when there are no active players on them for say, more than 30 minutes. This should work for every server (or at least the ones specified). Minecraft seems to always be a point of contention, but Minecraft isn't the only server that uses close to 5GB of ram with no users.

### Describe the solution you'd like.

Should the "bandwidth" not be available, I would like to ask if a bounty can be created for this. This is an issue and feature that's been requested several times in the discord and in various parts of the internet (including a closed feature request in this very github). The users want this and have been asking for this for a while. 

Should you not want to do a bounty either, would you mind please letting me know, approximately, what coding skills would be needed and how much you would charge for a bounty to get this task done if we were to utilize an outside bounty source (like bountysource.com)? Thank you!

**Comment by Rezanans-wow** [comment]
> when there are no active players on them for say, more than 30 minutes

So i think its not possible starting with this. As I know pterodactyl cant know (or get) the number of players on the server, because it does not have this option and as I know will not have. (I mean now only about do hibernation at 'no players at server in some time')

**Comment by Jcodeerd** [comment]
I don't think it's really necessary to add game-specific stuff to the panel. Pterodactyl is made for more games than just Minecraft. You might be able to do what you want with an egg.

**Comment by theraw** [comment]
> > when there are no active players on them for say, more than 30 minutes
> 
> So i think its not possible starting with this. As I know pterodactyl cant know (or get) the number of players on the server, because it does not have this option and as I know will not have. (I mean now only about do hibernation at 'no players at server in some time')

well this can't be done only in games query level its also possible to be done on network level. the network traffic that you can clearly see from dashboard and clearly ptero can access and monitor that. which is a start.

and i believe network level is only way if or idk well ppl can run many things on ptero who would add that many types of game queries. way easier to simply monitor container traffic. like may rest in peace heroku.

while putting a container to sleep may be way easier, to wake it up looks a bit more complicated and only way i personally see possible ATM is running a reverse proxy thing on host and waking up the container when new request is made to port.
another *problem* that i see is how this may affect game server ranking on gametracker while server is offline.

i however have seen big projects using GO (same language which wings is made) being used for reverse proxy tcp/udp and can handle some good traffic throughput even filter it for firewall purposes and i believe it could be possible to use wings for this however i don't know golang 🙄

would've loved if wings was in some different lang that i knew could've done so much with it i see wings as bit limiting of dockers functionality. for my purposes.

**Comment by BolverBlitz** [comment]
Since you mentioned minecraft spesificly i can recommend our [MSH](https://github.com/gekware/minecraft-server-hibernation) project. We made some eggs to get you started in no time. There are quite some "free" hosters who use it very activly and we got great feedback.
It got 2 modes, 1 will stop the server saving CPU and memory, the other will suspend the process and based on your OS config it will be moved into swap (freeing memory) while still provide startups within seconds for almost any size minecraft server, however this needs some server spezific tuning to work well.

The general request might be rather complicated because a lot of programs does not like beeing suspended for longer (nither does minecraft without config or trickery)
